api/src/main/java/org/apache/iceberg/expressions/AggregateEvaluator.java

```diff
@@ -0,0 +1,121 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.expressions.BoundAggregate.Aggregator;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+
+/**
+ * A class for evaluating aggregates. It evaluates each of the aggregates and updates the aggregated
+ * value. The final aggregated result can be returned by {@link #result()}.
+ */
+public class AggregateEvaluator {
+
+  public static AggregateEvaluator create(Schema schema, List<Expression> aggregates) {
+    return create(schema.asStruct(), aggregates);
+  }
+
+  private static AggregateEvaluator create(Types.StructType struct, List<Expression> aggregates) {
+    List<BoundAggregate<?, ?>> boundAggregates =
+        aggregates.stream()
+            .map(expr -> Binder.bind(struct, expr))
+            .map(bound -> (BoundAggregate<?, ?>) bound)
+            .collect(Collectors.toList());
+
+    return new AggregateEvaluator(boundAggregates);
+  }
+
+  private final List<Aggregator<?>> aggregators;
+  private final Types.StructType resultType;
+  private final List<BoundAggregate<?, ?>> aggregates;
+
+  private AggregateEvaluator(List<BoundAggregate<?, ?>> aggregates) {
+    ImmutableList.Builder<Aggregator<?>> aggregatorsBuilder = ImmutableList.builder();
+    List<Types.NestedField> resultFields = Lists.newArrayList();
+
+    for (int pos = 0; pos < aggregates.size(); pos += 1) {
+      BoundAggregate<?, ?> aggregate = aggregates.get(pos);
+      aggregatorsBuilder.add(aggregate.newAggregator());
+      resultFields.add(Types.NestedField.optional(pos, aggregate.describe(), aggregate.type()));
+    }
+
+    this.aggregators = aggregatorsBuilder.build();
+    this.resultType = Types.StructType.of(resultFields);
+    this.aggregates = aggregates;
+  }
+
+  public void update(StructLike struct) {
+    for (Aggregator<?> aggregator : aggregators) {
+      aggregator.update(struct);
+    }
+  }
+
+  public boolean update(DataFile file) {
+    for (Aggregator<?> aggregator : aggregators) {
+      aggregator.update(file);
+      if (!aggregator.hasValue()) {
```

Comment:
This should always call all aggregators. Each aggregator will stop aggregation independently if there is a missing value, then it is the responsibility of the caller to determine whether the aggregation can be used by checking if any aggregate has no value. The caller may still choose to get the aggregates for which there are values, so this should always return what it can. If there is no value, it should return `null`.